moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp

```diff
@@ -381,6 +381,11 @@ class MoveGroupInterfaceWrapper : protected py_bindings_tools::ROScppInitializer
     return getName().c_str();
   }
 
+  const char* getPlannerIdCStr() const
+  {
+    return getPlannerId().c_str();
+  }
+
```

Comment:
Why do you need the additional wrapper?

[It seems that returning std::string should work just fine](https://stackoverflow.com/questions/44827957/using-stdstring-as-return-type-in-boost-python-exported-function)

Reply:
I thought the same but I got some compilation errors when I tried.
When looking into it I saw that this is how the other methods returning strings (i.e. getName or getEndEffectorLink) were handling it, so I took the same approach.

Reply:
> I thought the same but I got some compilation errors when I tried.

You do not by any chance have the error message at hand?

It would be nice to see this cumbersome wrapping removed for the other methods as well.

Reply:
Hi @v4hn, I do have it. The error made me think that the python bindings didn't have a way to convert the return type from std::string:
```
                 from /home/gerard/code/catkin_ws/src/moveit/moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:38:
/usr/include/boost/python/detail/invoke.hpp: In instantiation of ‘PyObject* boost::python::detail::invoke(boost::python::detail::invoke_tag_<false, true>, const RC&, F&, TC&) [with RC = boost::python::detail::specify_a_return_value_policy_to_wrap_functions_returning<const std::__cxx11::basic_string<char>&>; F = const std::__cxx11::basic_string<char>& (moveit::planning_interface::MoveGroupInterface::*)() const; TC = boost::python::arg_from_python<moveit::planning_interface::MoveGroupInterfaceWrapper&>; PyObject = _object]’:
/usr/include/boost/python/detail/caller.hpp:216:46:   required from ‘PyObject* boost::python::detail::caller_arity<1>::impl<F, Policies, Sig>::operator()(PyObject*, PyObject*) [with F = const std::__cxx11::basic_string<char>& (moveit::planning_interface::MoveGroupInterface::*)() const; Policies = boost::python::default_call_policies; Sig = boost::mpl::vector2<const std::__cxx11::basic_string<char>&, moveit::planning_interface::MoveGroupInterfaceWrapper&>; PyObject = _object]’
/usr/include/boost/python/object/py_function.hpp:38:33:   required from ‘PyObject* boost::python::objects::caller_py_function_impl<Caller>::operator()(PyObject*, PyObject*) [with Caller = boost::python::detail::caller<const std::__cxx11::basic_string<char>& (moveit::planning_interface::MoveGroupInterface::*)() const, boost::python::default_call_policies, boost::mpl::vector2<const std::__cxx11::basic_string<char>&, moveit::planning_interface::MoveGroupInterfaceWrapper&> >; PyObject = _object]’
/home/gerard/code/catkin_ws/src/moveit_gcc/moveit/moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:721:1:   required from here
/usr/include/boost/python/detail/invoke.hpp:86:14: error: no match for call to ‘(const boost::python::detail::specify_a_return_value_policy_to_wrap_functions_returning<const std::__cxx11::basic_string<char>&>) (const std::__cxx11::basic_string<char>&)’
     return rc( (tc().*f)(BOOST_PP_ENUM_BINARY_PARAMS_Z(1, N, ac, () BOOST_PP_INTERCEPT)) );
            ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                 from /home/gerard/code/catkin_ws/src/moveit/moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:38:
/usr/include/boost/python/detail/caller.hpp: In instantiation of ‘static const PyTypeObject* boost::python::detail::converter_target_type<ResultConverter>::get_pytype() [with ResultConverter = boost::python::detail::specify_a_return_value_policy_to_wrap_functions_returning<const std::__cxx11::basic_string<char>&>; PyTypeObject = _typeobject]’:
/usr/include/boost/python/detail/caller.hpp:238:19:   required from ‘static boost::python::detail::py_func_sig_info boost::python::detail::caller_arity<1>::impl<F, Policies, Sig>::signature() [with F = const std::__cxx11::basic_string<char>& (moveit::planning_interface::MoveGroupInterface::*)() const; Policies = boost::python::default_call_policies; Sig = boost::mpl::vector2<const std::__cxx11::basic_string<char>&, moveit::planning_interface::MoveGroupInterfaceWrapper&>]’
/usr/include/boost/python/object/py_function.hpp:48:35:   required from ‘boost::python::detail::py_func_sig_info boost::python::objects::caller_py_function_impl<Caller>::signature() const [with Caller = boost::python::detail::caller<const std::__cxx11::basic_string<char>& (moveit::planning_interface::MoveGroupInterface::*)() const, boost::python::default_call_policies, boost::mpl::vector2<const std::__cxx11::basic_string<char>&, moveit::planning_interface::MoveGroupInterfaceWrapper&> >]’
/home/gerard/code/catkin_ws/src/moveit_gcc/moveit/moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:721:1:   required from here
/usr/include/boost/python/detail/caller.hpp:100:98: error: ‘struct boost::python::detail::specify_a_return_value_policy_to_wrap_functions_returning<const std::__cxx11::basic_string<char>&>’ has no member named ‘get_pytype’
         return create_result_converter((PyObject*)0, (ResultConverter *)0, (ResultConverter *)0).get_pytype();
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
```

After looking into int in more detail, I think the issue may be with the return type of all these methods returning strings being `const std::string&`. Boost python doesn't seem to like the reference in the return type. I just tried it changing the `MoveGroupInterface::getPlannerId()` method to return a `const std::string` instead and it compiles just fine. 

I guess there's some reason (efficiency?) for the MoveGroupInterface to return the strings as const refs, so we can also get rid of the `c_str()` making the wrapper methods return the `std::string` instead of `const char*` (although `MoveGroupInterface::getDefaultPlannerId` returns a plain string by value).

If you want I can check all the string returning methods, fix them in either way and add it to this PR before it gets merged (or into a new one if that's better).

Reply:
Argh, the const ref is a problem indeed. There might be some bp magic that could use the return types by their actual value (ignoring the ref), but that's quite a bit deeper than what I hoped for.

Although it is somewhat ridiculous to optimize the MGI on the level of const refs, when it's actually a high-delay adapter that uses ROS communication, that's what we have right now.
In this context, the current approach to extract the `c_str()` from the const ref is probably the most efficient one, so feel free to leave it as-is.

Thank you for following up on this!

Reply:
Sound good! Glad I was able to help a bit :-)